opnsense/plugins#4402: configure custom permissions on caddy.sock

g-a-c opened the pull request:
Fixes: opnsense/plugins#4402

When Caddy is given a Unix socket for its admin interface, and no permissions are supplied, the permissions become `0200` by default ([link](https://github.com/caddyserver/caddy/blob/v2.8.4/internal/sockets.go#L34-L55)).

As part of applying changes, OPNsense ends up running a script which results in the owner of `/var/run/caddy/caddy.sock` becoming `root`, thus when Caddy is running under the `www` user it can no longer perform admin operations like stop/restart.

This proposed change would provide custom permissions to Caddy, which is already catered for in the upstream code, to make sure that no matter who the **owner** of `caddy.sock` is, the `www` **group** can write to this socket. This way, whether Caddy is running under the `root` user or the `www` user, it can still use the admin API via this Unix socket.

fichtner commented on www/caddy/src/opnsense/service/templates/OPNsense/Caddy/rc.conf.d/caddy:
```diff
@@ -3,6 +3,7 @@
 {% set generalSettings = helpers.getNodeByTag('Pischem.caddy.general') %}
 {% if generalSettings.enabled|default("0") == "1" %}
 caddy_enable="YES"
+caddy_admin="unix//var/run/caddy/caddy.sock|0220"
```
for a service being able to drop privs and FreeBSD ports already overriding CADDY_ADMIN I think this should actually be the FreeBSD default.

fichtner replied:
For reference https://github.com/opnsense/ports/blob/master/www/caddy/files/caddy.in#L68

g-a-c replied:
> for a service being able to drop privs and FreeBSD ports already overriding CADDY_ADMIN I think this should actually be the FreeBSD default.

Perhaps - but I suspect upstream FreeBSD doesn't see the same issue because they don't have `setup.sh` changing ownership beneath them, which would mean the default `0200` permissions that Caddy sets on creating the socket would be valid for the lifetime of the process.

fichtner replied:
The fact it's relevant for us means caddy can be susceptible to the issue and FreeBSD port setting the socket over `localhost:2019` just means improving the situation in FreeBSD port sounds like a win-win.

After a bit of back and forth I just want to say your suggestion is very good and I don't see the harm in moving it to FreeBSD ports too.

fichtner commented on www/caddy/Makefile:
```diff
@@ -1,5 +1,5 @@
 PLUGIN_NAME=		caddy
-PLUGIN_VERSION=		1.7.5
+PLUGIN_VERSION=		1.7.6
 PLUGIN_REVISION=	1
```
please remove revision

fichtner commented:
Merged, thanks!

g-a-c commented:
> I've recently installed caddy and it didn't seem to start because it couldn't get a lock. In researching what could be happening I deleted /var/run/caddy - including the log.sock. I was expecting caddy to recreate the socket again when starting, but it doesn't. I haven't seen any scripts where caddy creates a socket. Would you have any suggestion how to recreate the socket? Uninstalling and installing doesn't work unfortunately.

Yes, Caddy should create the log socket when it starts up - it's part of the config in `/usr/local/etc/caddy/Caddyfile`. What are the settings you have Caddy using? Is it configured to drop privileges to `www` on startup? Can you explain a bit more about exactly what you did (i.e. installed Caddy and it didn't work immediately, installed Caddy and changed something then it stopped working?)

mdbraber commented:
I'm not sure exactly how I got it working again, but I thought it might have something to with syslog-ng, so I disabled that and then tried starting caddy. I can't exactly reproduce the steps, but after some stopping/starting syslog-ng and caddy it started working again!
